Feature/dissociacion wc

bosonie opened the pull request:
A wokchain to calculate the total energy of a diatomic molecule as a function of the distance between the two atoms.
Built using the EoS workchain as a footprint.
Adds also the CLI command and some tests.

bosonie commented:
@sphuber I would really appreciate an opinion about the way to return the results, at the moment I just return an array Energy vs Distance.

bosonie commented:
@sphuber, on a second though, probably it is better to include the support on the "magnetization_per_site" option in another PR. Could you then please review this one when you have time?

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+        spec.input('molecule', valid_type=orm.StructureData, validator=validate_molecule, help='The input molecule.')
+        spec.input('distances', valid_type=orm.List, required=False, validator=validate_distances,
+            help='The list of dinstances in Ang at which the total energy of the molecule should be computed.')
+        spec.input('distances_count', valid_type=orm.Int, default=lambda: orm.Int(20),
+            validator=validate_distances_count,
+            help='The number of points to compute in the dissociation curve.')
+        spec.input('distance_min', valid_type=orm.Float, default=lambda: orm.Float(0.5),
+            validator=validate_distance_min,
+            help='The minimum tested dinstance in Ang.')
+        spec.input('distance_max', valid_type=orm.Float, default=lambda: orm.Float(3),
+            validator=validate_distance_max,
+            help='The maximum tested distance in Ang.')
```
Wondering if we should specify the default here. If someone specifies an explicit list of `distances` then the defaults will still be added as nodes. Notice now that the same is happening for the `EosWorkChain` so maybe ok to leave for now but maybe open a low priority issue to maybe find a better solution. Downside of not having the defaults here is that they cannot be introspected through the spec explicitly

sphuber replied:
```suggestion
            help='The maximum tested distance in Ångstrom.')
```

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+        super().define(spec)
+        spec.input('molecule', valid_type=orm.StructureData, validator=validate_molecule, help='The input molecule.')
+        spec.input('distances', valid_type=orm.List, required=False, validator=validate_distances,
+            help='The list of dinstances in Ang at which the total energy of the molecule should be computed.')
```
```suggestion
            help='The list of distances in Ångstrom at which the total energy of the molecule should be computed.')
```

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+            help='The number of points to compute in the dissociation curve.')
+        spec.input('distance_min', valid_type=orm.Float, default=lambda: orm.Float(0.5),
+            validator=validate_distance_min,
+            help='The minimum tested dinstance in Ang.')
```
```suggestion
            help='The minimum tested distance in Ångstrom.')
```

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+            help='The type of relaxation to perform.')
+        spec.input('generator_inputs.spin_type', valid_type=SpinType, required=False, non_db=True,
+            help='The type of spin for the calculation.')
+        spec.input('generator_inputs.electronic_type', valid_type=ElectronicType, required=False, non_db=True,
+            help='The type of electronics (insulator/metal) for the calculation.')
```
Does this input make sense for a molecule?

bosonie replied:
It does not, but I now remember that the default is `METAL`. Any drawback in using metal options for molecules?

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+            return self.inputs.distances
+
+        count = self.inputs.distances_count.value
+        maxim = self.inputs.distance_max.value
+        minim = self.inputs.distance_min.value
```
I think you use `maxim` and `minim` because it makes it line up with `count` but I think using `maximum` and `minimum` is better for readability here. `max` and `min` would have been good too but that shadows a built-in

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+        if any([not child.is_finished_ok for child in self.ctx.children]):
+            return self.exit_codes.ERROR_SUB_PROCESS_FAILED.format(cls=self.inputs.sub_process_class)  # pylint: disable=no-member
+
+        self.report('distance (ang),  energy (eV)')
```
Do we really need to report the values here? Is it useful to have in the log?

bosonie replied:
We do it for the eos, so why not here?

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+            cls.run_dissociation,
+            cls.inspect_results,
+        )
+        spec.output('curve_data', valid_type=orm.ArrayData,
+            help='The computed total energy (and eventually spin) versus distance.')
```
Why don't we keep the same idea as for the EOS workchain and do something like the following:
```
spec.output_namespace('distances', valid_type=Float)
spec.output_namespace('total_energies', valid_type=Float)
```
In this way, there is no need to create new output nodes, you are simply returning existing inputs/outputs

bosonie replied:
I was just worried that it would return too many outputs. To have a decent curve at least 20 distances are needed. Isn't it a concern the amount of outputs?

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,216 @@
+
+        self.report('distance (ang),  energy (eV)')
+
+        collectwcinfo = {}
+        for index, child in enumerate(self.ctx.children):
+            dist = self.get_distances()[index].value
+            collectwcinfo[f'd{dist}'.replace('.', '_')] = child.outputs.total_energy
+            self.report(f'{self.get_distances()[index].value}, {child.outputs.total_energy.value}')
+
```
See my comment on the output spec on how I think we should return the outputs. I don't think we need a wrapping `ArrayData` but simply add existing nodes in a namespace like the EOS. Could here becomes a lot simpler as well then

sphuber commented on tests/workflows/dissociation_curve/test_workchain_diss_curve.py:
```diff
@@ -0,0 +1,88 @@
+def ctx():
+    """Return the context for a port validator."""
+    return None
+
```
I don't think there is a test for `validate_inputs` which I think is quite important. It should test that it raises if neither `distances` nor all of the other three are defined, and what should happen if both sets are defined. As said in another comment, since the `distance_max` etc. define defaults, they will _always_ be passed, even if `distances` is explicitly defined and `distances` should then take the precedenc

bosonie replied:
I'm noticing now that the test is missing also for the eos... And the not tested `validate_inputs` of eos contains a bug! I will fix it here if you don't mind!

sphuber reviewed (changes requested):
Thanks @bosonie , have a few suggestions

bosonie commented on aiida_common_workflows/cli/launch.py:
```diff
@@ -228,6 +228,105 @@ def cmd_eos(
+@click.argument('plugin', type=types.LazyChoice(functools.partial(get_workflow_entry_point_names, 'relax', True)))
+@options.STRUCTURE(default='H2')
+@options.PROTOCOL(type=click.Choice(['fast', 'moderate', 'precise']), default='fast')
+@options.SPIN_TYPE()
```
Yes we do. I put it in my last commit. The test case of the hydrogen molecule will, in fact, require the two hydrogen atoms to have opposite spin.

bosonie commented on aiida_common_workflows/cli/launch.py:
```diff
@@ -228,6 +228,105 @@ def cmd_eos(
+    """Compute the dissociation curve of a diatomic molecule using the common relax workflow (in single point
+    relaxation mode - i.e. RelaxType.NONE).
+
+    The command will automatically try to find and load the codes that are required by the plugin workflow. If no code
+    is installed for at least one of the calculation engines, the command will fail. Use the `--show-engine` flag to
```
You are perfectly right!

bosonie commented:
@sphuber I addressed all your points accept the output format. I'm a bit worried about the too many outputs. Can you convince me that it is not bad to have 50 otputs with one single `float` on each one of them?
Also I think we should address the point you rise about `ElectronicType` for molecule tomorrow in the meeting.

sphuber commented:
> Can you convince me that it is not bad to have 50 outputs containing one single `float`?

My argument would be that you aren't creating an _additional_ 50 output nodes, but merely the links. The nodes already exists. The distance `Float` nodes already exist as inputs to the `set_distance` calcfunction and the `total_energy` nodes are the output of the `CommonRelaxWorkChain`. Adding the links is cheap and should not be a problem. We have databases with tens of millions of links and that doesn't cause an issue so far.

Note that your current solution _also_ adds a lot of links but in addition to that more nodes. You pass all `total_energy` output nodes to the `get_curve_data` function, and one link is created for each of them. In addition, you add more nodes to the database, one for the calcfunction itself and one for the `ArrayData`.

bosonie commented:
@sphuber I change the outputs as you suggested. Ready for a new review.

sphuber commented on aiida_common_workflows/workflows/dissociation.py:
```diff
@@ -0,0 +1,200 @@
+        if any(key not in value for key in ['distances_count', 'distance_min', 'distance_max']):
+            return 'neither `distances` nor the `distances_count`, `distance_min`, and `distance_max` set were defined.'
+    if 'distance_min' in value:
+        if value['distance_min'] >= value['distance_max']:
```
what if `distance_max` is not defined here?

bosonie replied:
You are right. It fails when `distance_min` and `distances` are defined, but not `distance_max`. a bit of a weird case but I'll change it in next commit

sphuber replied:
The change is all good, I am just wondering now if it isn't a bit weird to allow an empty list for `distances`. Anyway, I will merge this now and then we can address this later

bosonie replied:
It is not allowed! There is the `validator_distances` that takes care of it later on!

sphuber reviewed (commented):
Thanks @bosonie . I think this is almost good to go. I just had one more question on one of the validators. Should we wait with merging until the discussion tomorrow so we can nail down the question on whether the `electronic_type` should be an input?

bosonie commented:
Please check if my last commit is ok @sphuber. I now test `distance_max` and `distance_min` only if the `distances` is not present, reflecting the fact that `distance_max` and `distance_min` are unused if `distances` is present.

I would merge this PR without waiting today's discussion. Most likely, if some modifications need to be done for `electronic_type`, they will go beyond this test case. Better to do it in another PR.
